{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: Matrix\n",
      "\n",
      "Loading required package: foreach\n",
      "\n",
      "Loaded glmnet 2.0-13\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#-------------------------- arguments ---------------------------#\n",
    "options(stringsAsFactors=F)\n",
    "library(glmnet)\n",
    "library(foreach)\n",
    "library(genio)\n",
    "library(Rcpp)\n",
    "\n",
    " "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'adjust_expr_updated.R'</li><li>'adjust_expr.R'</li><li>'adjust_expression_gtex8_debug.R'</li><li>'adjust_expression_gtex8_parallel.R'</li><li>'adjust_expression_gtex8.R'</li><li>'check_adjusted_expression.ipynb'</li><li>'cmds_gen.R'</li><li>'correlation_snp_exp_mapping_MetaXcan.ipynb'</li><li>'correlation_snp_exp.ipynb'</li><li>'ctimp_working_version-Copy1.ipynb'</li><li>'ctimp_working_version.ipynb'</li><li>'EDA.ipynb'</li><li>'expression_adjustment_miRNA.ipynb'</li><li>'expression_adjustment_output.ipynb'</li><li>'expression_adjustment.ipynb'</li><li>'expression_check.ipynb'</li><li>'extract_dosage_info.ipynb'</li><li>'extract_raw_expression_v0.R'</li><li>'extract_raw_expression-Copy1.R'</li><li>'extract_raw_expression.R'</li><li>'gene_id_prep.ipynb'</li><li>'gene_info.ipynb'</li><li>'gtex_v7_Adipose_Subcutaneous_imputed_eur_covariances.txt.gz'</li><li>'predictdb_overall.ipynb'</li><li>'predictdb_summary-Copy1.ipynb'</li><li>'predictdb_summary.ipynb'</li><li>'predictdb_weight_mapping.ipynb'</li><li>'predictdb-Copy2.ipynb'</li><li>'predictdb.ipynb'</li><li>'snp_dosage_info-Copy1.ipynb'</li><li>'snp_dosage_info.ipynb'</li><li>'summarize_snp_direction.R'</li><li>'test_c_function.ipynb'</li><li>'test_ctimp_faster-Copy1.ipynb'</li><li>'test_ctimp_faster-Copy2.ipynb'</li><li>'test_ctimp_faster.ipynb'</li><li>'test_ctimp_singleest.ipynb'</li><li>'test_ctimp_xdimension-Copy1.ipynb'</li><li>'test_ctimp_xdimension.ipynb'</li><li>'test_ctimp-Copy1.ipynb'</li><li>'test_ctimp-Copy1.ipynb.invalid'</li><li>'test_ctimp-Copy2.ipynb'</li><li>'test_ctimp-Copy3.ipynb'</li><li>'test_ctimp.ipynb'</li><li>'Untitled.ipynb'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'adjust\\_expr\\_updated.R'\n",
       "\\item 'adjust\\_expr.R'\n",
       "\\item 'adjust\\_expression\\_gtex8\\_debug.R'\n",
       "\\item 'adjust\\_expression\\_gtex8\\_parallel.R'\n",
       "\\item 'adjust\\_expression\\_gtex8.R'\n",
       "\\item 'check\\_adjusted\\_expression.ipynb'\n",
       "\\item 'cmds\\_gen.R'\n",
       "\\item 'correlation\\_snp\\_exp\\_mapping\\_MetaXcan.ipynb'\n",
       "\\item 'correlation\\_snp\\_exp.ipynb'\n",
       "\\item 'ctimp\\_working\\_version-Copy1.ipynb'\n",
       "\\item 'ctimp\\_working\\_version.ipynb'\n",
       "\\item 'EDA.ipynb'\n",
       "\\item 'expression\\_adjustment\\_miRNA.ipynb'\n",
       "\\item 'expression\\_adjustment\\_output.ipynb'\n",
       "\\item 'expression\\_adjustment.ipynb'\n",
       "\\item 'expression\\_check.ipynb'\n",
       "\\item 'extract\\_dosage\\_info.ipynb'\n",
       "\\item 'extract\\_raw\\_expression\\_v0.R'\n",
       "\\item 'extract\\_raw\\_expression-Copy1.R'\n",
       "\\item 'extract\\_raw\\_expression.R'\n",
       "\\item 'gene\\_id\\_prep.ipynb'\n",
       "\\item 'gene\\_info.ipynb'\n",
       "\\item 'gtex\\_v7\\_Adipose\\_Subcutaneous\\_imputed\\_eur\\_covariances.txt.gz'\n",
       "\\item 'predictdb\\_overall.ipynb'\n",
       "\\item 'predictdb\\_summary-Copy1.ipynb'\n",
       "\\item 'predictdb\\_summary.ipynb'\n",
       "\\item 'predictdb\\_weight\\_mapping.ipynb'\n",
       "\\item 'predictdb-Copy2.ipynb'\n",
       "\\item 'predictdb.ipynb'\n",
       "\\item 'snp\\_dosage\\_info-Copy1.ipynb'\n",
       "\\item 'snp\\_dosage\\_info.ipynb'\n",
       "\\item 'summarize\\_snp\\_direction.R'\n",
       "\\item 'test\\_c\\_function.ipynb'\n",
       "\\item 'test\\_ctimp\\_faster-Copy1.ipynb'\n",
       "\\item 'test\\_ctimp\\_faster-Copy2.ipynb'\n",
       "\\item 'test\\_ctimp\\_faster.ipynb'\n",
       "\\item 'test\\_ctimp\\_singleest.ipynb'\n",
       "\\item 'test\\_ctimp\\_xdimension-Copy1.ipynb'\n",
       "\\item 'test\\_ctimp\\_xdimension.ipynb'\n",
       "\\item 'test\\_ctimp-Copy1.ipynb'\n",
       "\\item 'test\\_ctimp-Copy1.ipynb.invalid'\n",
       "\\item 'test\\_ctimp-Copy2.ipynb'\n",
       "\\item 'test\\_ctimp-Copy3.ipynb'\n",
       "\\item 'test\\_ctimp.ipynb'\n",
       "\\item 'Untitled.ipynb'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'adjust_expr_updated.R'\n",
       "2. 'adjust_expr.R'\n",
       "3. 'adjust_expression_gtex8_debug.R'\n",
       "4. 'adjust_expression_gtex8_parallel.R'\n",
       "5. 'adjust_expression_gtex8.R'\n",
       "6. 'check_adjusted_expression.ipynb'\n",
       "7. 'cmds_gen.R'\n",
       "8. 'correlation_snp_exp_mapping_MetaXcan.ipynb'\n",
       "9. 'correlation_snp_exp.ipynb'\n",
       "10. 'ctimp_working_version-Copy1.ipynb'\n",
       "11. 'ctimp_working_version.ipynb'\n",
       "12. 'EDA.ipynb'\n",
       "13. 'expression_adjustment_miRNA.ipynb'\n",
       "14. 'expression_adjustment_output.ipynb'\n",
       "15. 'expression_adjustment.ipynb'\n",
       "16. 'expression_check.ipynb'\n",
       "17. 'extract_dosage_info.ipynb'\n",
       "18. 'extract_raw_expression_v0.R'\n",
       "19. 'extract_raw_expression-Copy1.R'\n",
       "20. 'extract_raw_expression.R'\n",
       "21. 'gene_id_prep.ipynb'\n",
       "22. 'gene_info.ipynb'\n",
       "23. 'gtex_v7_Adipose_Subcutaneous_imputed_eur_covariances.txt.gz'\n",
       "24. 'predictdb_overall.ipynb'\n",
       "25. 'predictdb_summary-Copy1.ipynb'\n",
       "26. 'predictdb_summary.ipynb'\n",
       "27. 'predictdb_weight_mapping.ipynb'\n",
       "28. 'predictdb-Copy2.ipynb'\n",
       "29. 'predictdb.ipynb'\n",
       "30. 'snp_dosage_info-Copy1.ipynb'\n",
       "31. 'snp_dosage_info.ipynb'\n",
       "32. 'summarize_snp_direction.R'\n",
       "33. 'test_c_function.ipynb'\n",
       "34. 'test_ctimp_faster-Copy1.ipynb'\n",
       "35. 'test_ctimp_faster-Copy2.ipynb'\n",
       "36. 'test_ctimp_faster.ipynb'\n",
       "37. 'test_ctimp_singleest.ipynb'\n",
       "38. 'test_ctimp_xdimension-Copy1.ipynb'\n",
       "39. 'test_ctimp_xdimension.ipynb'\n",
       "40. 'test_ctimp-Copy1.ipynb'\n",
       "41. 'test_ctimp-Copy1.ipynb.invalid'\n",
       "42. 'test_ctimp-Copy2.ipynb'\n",
       "43. 'test_ctimp-Copy3.ipynb'\n",
       "44. 'test_ctimp.ipynb'\n",
       "45. 'Untitled.ipynb'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       " [1] \"adjust_expr_updated.R\"                                      \n",
       " [2] \"adjust_expr.R\"                                              \n",
       " [3] \"adjust_expression_gtex8_debug.R\"                            \n",
       " [4] \"adjust_expression_gtex8_parallel.R\"                         \n",
       " [5] \"adjust_expression_gtex8.R\"                                  \n",
       " [6] \"check_adjusted_expression.ipynb\"                            \n",
       " [7] \"cmds_gen.R\"                                                 \n",
       " [8] \"correlation_snp_exp_mapping_MetaXcan.ipynb\"                 \n",
       " [9] \"correlation_snp_exp.ipynb\"                                  \n",
       "[10] \"ctimp_working_version-Copy1.ipynb\"                          \n",
       "[11] \"ctimp_working_version.ipynb\"                                \n",
       "[12] \"EDA.ipynb\"                                                  \n",
       "[13] \"expression_adjustment_miRNA.ipynb\"                          \n",
       "[14] \"expression_adjustment_output.ipynb\"                         \n",
       "[15] \"expression_adjustment.ipynb\"                                \n",
       "[16] \"expression_check.ipynb\"                                     \n",
       "[17] \"extract_dosage_info.ipynb\"                                  \n",
       "[18] \"extract_raw_expression_v0.R\"                                \n",
       "[19] \"extract_raw_expression-Copy1.R\"                             \n",
       "[20] \"extract_raw_expression.R\"                                   \n",
       "[21] \"gene_id_prep.ipynb\"                                         \n",
       "[22] \"gene_info.ipynb\"                                            \n",
       "[23] \"gtex_v7_Adipose_Subcutaneous_imputed_eur_covariances.txt.gz\"\n",
       "[24] \"predictdb_overall.ipynb\"                                    \n",
       "[25] \"predictdb_summary-Copy1.ipynb\"                              \n",
       "[26] \"predictdb_summary.ipynb\"                                    \n",
       "[27] \"predictdb_weight_mapping.ipynb\"                             \n",
       "[28] \"predictdb-Copy2.ipynb\"                                      \n",
       "[29] \"predictdb.ipynb\"                                            \n",
       "[30] \"snp_dosage_info-Copy1.ipynb\"                                \n",
       "[31] \"snp_dosage_info.ipynb\"                                      \n",
       "[32] \"summarize_snp_direction.R\"                                  \n",
       "[33] \"test_c_function.ipynb\"                                      \n",
       "[34] \"test_ctimp_faster-Copy1.ipynb\"                              \n",
       "[35] \"test_ctimp_faster-Copy2.ipynb\"                              \n",
       "[36] \"test_ctimp_faster.ipynb\"                                    \n",
       "[37] \"test_ctimp_singleest.ipynb\"                                 \n",
       "[38] \"test_ctimp_xdimension-Copy1.ipynb\"                          \n",
       "[39] \"test_ctimp_xdimension.ipynb\"                                \n",
       "[40] \"test_ctimp-Copy1.ipynb\"                                     \n",
       "[41] \"test_ctimp-Copy1.ipynb.invalid\"                             \n",
       "[42] \"test_ctimp-Copy2.ipynb\"                                     \n",
       "[43] \"test_ctimp-Copy3.ipynb\"                                     \n",
       "[44] \"test_ctimp.ipynb\"                                           \n",
       "[45] \"Untitled.ipynb\"                                             "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "list.files()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "#-------------------------- data import ---------------------------#  \n",
    "\n",
    "info_list <- list()\n",
    "genotype_dir <- \"/ysm-gpfs/project/wl382/GTEx_v8/genotype/cis_loc/\"\n",
    "\n",
    "count = 0\n",
    "for (chr in 1:22) {\n",
    "    IRdisplay::display_html(paste0(\"INFO: chr \", chr))\n",
    "    info_list[[chr]] <- list()\n",
    "    gtex_dir <- \"/gpfs/loomis/project/zhao/zy92/GTEX/\" \n",
    "    chr_str <- paste0(\"chr\", chr, \"/\")\n",
    "    #gene_vec <- list.files(paste0(genotype_dir, chr_str))\n",
    "    gene_vec = dir(paste0(gtex_dir, \"adjusted_expr/chr\", chr))\n",
    "    \n",
    "    for (gene_id in gene_vec) {\n",
    "        count = count + 1\n",
    "        if (count %% 50 == 0) {IRdisplay::display_html(paste0(\"INFO: chr \", chr, \" gene \", count))}\n",
    "        dose_path <- paste0(genotype_dir, chr_str, gene_id, \"/\", gene_id)\n",
    "        Yt <- dir(paste0(gtex_dir, \"adjusted_expr/\", chr_str, gene_id, \"/\"))\n",
    "        P <- length(Yt)\n",
    "        Y = list()\n",
    "\n",
    "        for(t in 1:P){\n",
    "            Y[[t]] = read.table(paste0(gtex_dir, \"adjusted_expr/chr\", chr, \"/\", gene_id, \"/\", Yt[t]), header=F)\n",
    "        }\n",
    "        \n",
    "        ssize = unlist(lapply(Y, nrow))\n",
    "        T_num = length(Yt)\n",
    "        info_list[[chr]][[gene_id]] <- list(T_num, ssize, Y)\n",
    "        }\n",
    "\n",
    "}\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "\n",
    "\n",
    "#-------------------------- functions ---------------------------#  \n",
    "\n",
    "grad_prep <- function(X, Y){\n",
    "\t## pre-calculate some metrics for gradient\n",
    "\tll = length(Y)\n",
    "\tP = ncol(X[[1]])\n",
    "\tXY = matrix(0,P,ll)\n",
    "\tfor(i in 1:ll){\n",
    "\t\tXY[,i] = t(X[[i]])%*%Y[[i]]/nrow(X[[i]])\n",
    "\t}\n",
    "\tXY\n",
    "}\n",
    "\n",
    "cv_helper <- function(N, fold){\n",
    "\tvalid_num = floor(N/fold)\n",
    "\tset.seed(123)\n",
    "\tperm = sample(1:N, size = N)\n",
    "\tidx1 = seq(1,N,valid_num)\n",
    "\tidx2 = c(idx1[-1]-1,N)\n",
    "\tlist(perm=perm, idx=cbind(idx1,idx2))\n",
    "}\n",
    "\n",
    "minmax_lambda <- function(lst){\n",
    "\tmax_lam = max(unlist(lapply(lst, function(x){max(x$lambda)})))\n",
    "\tmin_lam = min(unlist(lapply(lst, function(x){min(x$lambda)})))\n",
    "\tc(min_lam, max_lam)\n",
    "}\n",
    "\n",
    "elastic_net_mse <- function(lst, X_tune, Y_tune, X_test, Y_test){\n",
    "\tP = length(lst)\n",
    "\tM = ncol(X_tune[[1]])\n",
    "\tlam_V = rep(0, P)\n",
    "\ttest_res = list()\n",
    "\ttest_beta = matrix(0, M, P)\n",
    "\tfor(t in 1:P){\n",
    "\t\tncv = length(lst[[t]]$lambda)\n",
    "\t\ttmp_mse = rep(0, ncv)\n",
    "\t\tfor(k in 1:ncv){\n",
    "\t\t\ttmp_mse[k] = mean((Y_tune[[t]] - X_tune[[t]]%*%lst[[t]]$glmnet.fit$beta[,k])^2)\n",
    "\t\t}\n",
    "\t\tss = which.min(tmp_mse)\n",
    "\t\ttest_beta[,t] = lst[[t]]$glmnet.fit$beta[,ss]\n",
    "\t\tlam_V[t] = lst[[t]]$lambda[ss]\n",
    "\t\tpredicted = X_test[[t]]%*%lst[[t]]$glmnet.fit$beta[,ss]\n",
    "\t\ttest_res[[t]] = cbind(Y_test[[t]], predicted)\n",
    "\t}\n",
    "\tlist(lam = lam_V, mse = test_res, est = test_beta)\n",
    "}\n",
    "\n",
    "multi_mse <- function(theta_est, X_test, Y_test){\n",
    "\tanswer = list()\n",
    "\tP = ncol(theta_est)\n",
    "\tfor(t in 1:P){\n",
    "\t\tpredicted = X_test[[t]]%*%theta_est[,t]\n",
    "\t\tanswer[[t]] = cbind(Y_test[[t]], predicted)\n",
    "\t}\n",
    "\tanswer\n",
    "}\n",
    "\n",
    "avg_perm <- function(mse_lst){\n",
    "\tfd = length(mse_lst)\n",
    "\tP = length(mse_lst[[1]])\n",
    "\trsq = mse = adj_mse = matrix(0, fd, P)\n",
    "\tfor(f in 1:fd){\n",
    "\t\tfor(t in 1:P){\n",
    "\t\t\trsq[f,t] = (cor(mse_lst[[f]][[t]])[1,2])^2\n",
    "\t\t\tmse[f,t] = mean((mse_lst[[f]][[t]][,1]-mse_lst[[f]][[t]][,2])^2)\n",
    "\t\t\tadj_mse[f,t] = mse[f,t]/var(mse_lst[[f]][[t]][,1])\n",
    "\t\t}\n",
    "\t}\n",
    "\tcbind(apply(rsq, 2, mean), apply(mse, 2, mean), apply(adj_mse, 2, mean))\n",
    "\n",
    "\t#list(rsq = apply(rsq, 2, mean), mse = apply(mse, 2, mean), adj_mse = apply(adj_mse, 2, mean))\n",
    "}\n",
    "\n",
    "glasso <- function(X, Y, X1, Y1, XX, XY, Xnorm, lambda1, lambda2, theta, stepsize = 1e-4, maxiter = 10, eps = 1e-4){\n",
    "\tbgt = Sys.time()\n",
    "\tM = nrow(XY)\n",
    "\tP = length(X)\n",
    "\tNN = unlist(lapply(X, nrow))\n",
    "\told_objV1 = 0\n",
    "\tfor(t in 1:P){\n",
    "\t\told_objV1 = old_objV1 + 1/2*mean((Y[[t]]-X[[t]]%*%theta[,t])^2)\n",
    "\t}\n",
    "\tIRdisplay::display_html(paste0(\"Training error: \", old_objV1, '\\n'))\n",
    "\told_objV2 = 0\n",
    "\tfor(t in 1:P){\n",
    "\t\told_objV2 = old_objV2 + 1/2*mean((Y1[[t]]-X1[[t]]%*%theta[,t])^2)\n",
    "\t}\n",
    "\tIRdisplay::display_html(paste0(\"Testing error: \", old_objV2, '\\n'))\n",
    "\tbeta_j_lasso = rep(0, P)\n",
    "\ttmp_XYj = 0\n",
    "\tif(!is.loaded(\"wrapper\")){\n",
    "\t\tdyn.load(\"/ysm-gpfs/pi/zhao/from_louise/yh367/GTEX/codes/glasso.so\")\n",
    "\t}\n",
    "    \n",
    "\tfor(i in 1:maxiter){\n",
    "\t\tbgt = Sys.time()\n",
    "\t\tres = .Call(\"wrapper\", XX, XY, theta, M, P, beta_j_lasso, lambda1, lambda2, Xnorm)\n",
    "\t\tedt = Sys.time()\n",
    "\t\t#print(edt-bgt)\n",
    "\t\tnew_objV1 = new_objV2 = 0\n",
    "\t\tfor(t in 1:P){\n",
    "\t\t\tnew_objV1 = new_objV1 + 1/2*mean((Y[[t]]-X[[t]]%*%theta[,t])^2)\n",
    "            #IRdisplay::display_html(sum(is.na(X[[t]])))\n",
    "            #IRdisplay::display_html(sum(is.na(Y[[t]])))\n",
    "\t\t}\n",
    "\t\tIRdisplay::display_html(paste0(\"Training error: \", new_objV1, '\\n'))\n",
    "\t\tfor(t in 1:P){\n",
    "\t\t\tnew_objV2 = new_objV2 + 1/2*mean((Y1[[t]]-X1[[t]]%*%theta[,t])^2)\n",
    "           # IRdisplay::display_html(sum(is.na(X[[t]])))\n",
    "            #IRdisplay::display_html(sum(is.na(Y[[t]])))\n",
    "\t\t}\n",
    "\t\tIRdisplay::display_html(paste0(\"Testing error: \", new_objV2, '\\n'))\n",
    "\t\tif(new_objV2 > old_objV2|new_objV1 > old_objV1){\n",
    "\t\t\tbreak\n",
    "\t\t}else{\n",
    "\t\t\told_objV2 = new_objV2\n",
    "\t\t}\n",
    "\t\tif(abs(new_objV1-old_objV1) < eps){\n",
    "\t\t\tbreak\n",
    "\t\t}else{\n",
    "\t\t\told_objV1 = new_objV1\n",
    "\t\t}\n",
    "\t}\n",
    "\t#edt = Sys.time()\n",
    "\t#print(edt-bgt)\n",
    "\tlist(est = theta, tune_err = new_objV2)\n",
    "}\n",
    "\n",
    "glasso_no_early_stopping <- function(X, Y, XX, XY, Xnorm, lambda1, lambda2, theta, stepsize = 1e-4, maxiter = 250, eps = 1e-3){\n",
    "\tM = nrow(XY)\n",
    "\tP = length(X)\n",
    "\tNN = unlist(lapply(X, nrow))\n",
    "\told_objV1 = 0\n",
    "\tfor(t in 1:P){\n",
    "\t\told_objV1 = old_objV1 + 1/2*mean((Y[[t]]-X[[t]]%*%theta[,t])^2)\n",
    "\t}\n",
    "\tIRdisplay::display_html(paste0(\"Training error: \", old_objV1, '\\n'))\n",
    "\tbeta_j_lasso = rep(0, P)\n",
    "\ttmp_XYj = 0\n",
    "\tif(!is.loaded(\"wrapper\")){\n",
    "\t\tdyn.load(\"/ysm-gpfs/pi/zhao/from_louise/yh367/GTEX/codes/glasso.so\")\n",
    "\t}\n",
    "\tfor(i in 1:maxiter){\n",
    "\t\tres = .Call(\"wrapper\", XX, XY, theta, M, P, beta_j_lasso, lambda1, lambda2, Xnorm)\n",
    "\t\tnew_objV1 = 0\n",
    "\t\tfor(t in 1:P){\n",
    "\t\t\tnew_objV1 = new_objV1 + 1/2*mean((Y[[t]]-X[[t]]%*%theta[,t])^2)\n",
    "\t\t}\n",
    "\t\tIRdisplay::display_html(paste0(\"Training error: \", new_objV1, '\\n'))\n",
    "\t\tif(abs(new_objV1-old_objV1) < eps|new_objV1 > old_objV1){\n",
    "\t\t\tbreak\n",
    "\t\t}else{\n",
    "\t\t\told_objV1 = new_objV1\n",
    "\t\t}\n",
    "\t}\n",
    "\tlist(est = theta, train_err = new_objV1)\n",
    "}\n",
    "\n",
    "extract_genotype <- function(gene_idx, chr_str, genotype_dir, gene_vec) {\n",
    "    gene_dir <- paste0(genotype_dir, chr_str, gene_vec[gene_idx], \"/\", gene_vec[gene_idx])\n",
    "    genotype_info <- read_plink(gene_dir)\n",
    "    \n",
    "    print(paste0(\"INFO: genotype matrix dimension:\", dim(genotype_info$X)[1], \" * \", dim(genotype_info$X)[2]))\n",
    "    return(genotype_info)\n",
    "}\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO: loading expression files..."
     ]
    }
   ],
   "source": [
    "# cv_config\n",
    "fold = 5\n",
    "cv_config = cv_helper(450, fold)\n",
    "\n",
    "# create dirs\n",
    "dir.create(paste0(output_dir, \"chr\", chr), showWarnings = FALSE)\n",
    "dir.create(paste0(output_dir, \"chr\", chr, \"/\", gene_id), showWarnings = FALSE)\n",
    "setwd(paste0(output_dir, \"chr\", chr, \"/\", gene_id))\n",
    "\n",
    "# expr files \n",
    "cat(\"INFO: loading expression files...\")\n",
    "Y = list()\n",
    "for(t in 1:P){\n",
    "    Y[[t]] = read.table(paste0(gtex_dir, \"adjusted_expr/chr\", chr, \"/\", g, \"/\", Yt[t]), header=F)\n",
    "}\n",
    "ssize = unlist(lapply(Y, nrow))\n",
    "T_num = length(Yt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# cv_config\n",
    "fold = 5\n",
    "cv_config = cv_helper(450, fold)\n",
    "\n",
    "# create dirs\n",
    "dir.create(paste0(output_dir, \"chr\", chr), showWarnings = FALSE)\n",
    "dir.create(paste0(output_dir, \"chr\", chr, \"/\", gene_id), showWarnings = FALSE)\n",
    "setwd(paste0(output_dir, \"chr\", chr, \"/\", gene_id))\n",
    "\n",
    "# expr files \n",
    "cat(\"INFO: loading expression files...\")\n",
    "Y = list()\n",
    "for(t in 1:P){\n",
    "    Y[[t]] = read.table(paste0(gtex_dir, \"adjusted_expr/chr\", chr, \"/\", g, \"/\", Yt[t]), header=F)\n",
    "}\n",
    "ssize = unlist(lapply(Y, nrow))\n",
    "T_num = length(Yt)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.5.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
